### ISSUE TITLE
Exception 0xc0000005 0x1 0x333249 0x6c7bec77 PC=0x6c7bec77 signal arrived during cgo execution

### ISSUE THREAD

**Issue by brunoqc** [veandco/go-sdl2#78]
I have this intermittent problem. Sometimes it crashes on launch.

```go
package main

import (
	"fmt"
	"runtime"

	"github.com/veandco/go-sdl2/sdl"
	"github.com/veandco/go-sdl2/sdl_ttf"
)

func main() {
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	ttf.Init()

	font, errOpen := ttf.OpenFont("Ubuntu-B.ttf", 300)
	if errOpen != nil {
		panic(errOpen)
	}
	defer font.Close()

	window, err := sdl.CreateWindow("test", sdl.WINDOWPOS_UNDEFINED, sdl.WINDOWPOS_UNDEFINED,
		800, 600, sdl.WINDOW_SHOWN)
	if err != nil {
		panic(err)
	}
	defer window.Destroy()

	surface := window.GetSurface()

	fontSurface := font.RenderText_Shaded("123", sdl.Color{R: 255, G: 165, B: 0, A: 255}, sdl.Color{R: 0, G: 0, B: 0, A: 255})
	rect := sdl.Rect{surface.W/2 - fontSurface.W/2, surface.H/2 - fontSurface.H/2, 800, 800}
	fontSurface.Blit(nil, surface, &rect)
	fontSurface.Free()

	var event sdl.Event
	running := true

	for running {
		for event = sdl.PollEvent(); event != nil; event = sdl.PollEvent() {
			switch t := event.(type) {
			case *sdl.QuitEvent:
				running = false
			case *sdl.MouseMotionEvent:
				fmt.Printf("[%d ms] MouseMotion\ttype:%d\tid:%d\tx:%d\ty:%d\txrel:%d\tyrel:%d\n",
					t.Timestamp, t.Type, t.Which, t.X, t.Y, t.XRel, t.YRel)
			case *sdl.MouseButtonEvent:
				fmt.Printf("[%d ms] MouseButton\ttype:%d\tid:%d\tx:%d\ty:%d\tbutton:%d\tstate:%d\n",
					t.Timestamp, t.Type, t.Which, t.X, t.Y, t.Button, t.State)
			case *sdl.MouseWheelEvent:
				fmt.Printf("[%d ms] MouseWheel\ttype:%d\tid:%d\tx:%d\ty:%d\n",
					t.Timestamp, t.Type, t.Which, t.X, t.Y)
			case *sdl.KeyUpEvent:
				fmt.Printf("[%d ms] Keyboard\ttype:%d\tsym:%c\tmodifiers:%d\tstate:%d\trepeat:%d\n",
					t.Timestamp, t.Type, t.Keysym.Sym, t.Keysym.Mod, t.State, t.Repeat)
			}
		}

		window.UpdateSurface()
		sdl.Delay(1000 / 30)
	}
}
```

```
Exception 0xc0000005 0x1 0x333249 0x6c7bec77
PC=0x6c7bec77
signal arrived during cgo execution

github.com/veandco/go-sdl2/sdl_ttf._Cfunc_SDL_free(0x8ce750)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/:44 +0x4c
github.com/veandco/go-sdl2/sdl_ttf.(*Font).RenderText_Shaded(0xc082034018, 0x517010, 0x3, 0xff000000ff00a5ff, 0x2829a0)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/sdl_ttf.go:110 +0x11c
main.main()
        D:/Dev/go/Tests/sdl2crash/sdl2.go:32 +0x24b

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        c:/go/src/runtime/asm_amd64.s:2232 +0x1
rax     0x8ce740
rbx     0x8
rcx     0x6c85fb60
rdx     0x333231
rdi     0xc082025d00
rsi     0x5bfa60
rbp     0x22feb0
rsp     0x22fdf0
r8      0x1c00bf0e2bb52254
r9      0x1c00bf0e2c420994
r10     0x8c0158
r11     0x280000
r12     0xff
r13     0x0
r14     0x0
r15     0x0
rip     0x6c7bec77
rflags  0x10293
cs      0x33
fs      0x53
gs      0x2b
```

**Comment by jackyb** [comment]
I can't reproduce the error but I put a lock inside RenderText_Shaded(). Could you check if it helps?

**Comment by brunoqc** [comment]
Do you mean like this?

```go
func (f *Font) RenderText_Shaded(text string, fg, bg sdl.Color) *sdl.Surface {
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	_text := C.CString(text)
	defer C.SDL_free(unsafe.Pointer(_text))
	_fg := C.SDL_Color{C.Uint8(fg.R), C.Uint8(fg.G), C.Uint8(fg.B), C.Uint8(fg.A)}
	_bg := C.SDL_Color{C.Uint8(bg.R), C.Uint8(bg.G), C.Uint8(bg.B), C.Uint8(bg.A)}
	surface := (*sdl.Surface)(unsafe.Pointer(C.TTF_RenderText_Shaded(f.f, _text, _fg, _bg)))
	return surface
}
```

**Comment by jackyb** [comment]
Yes, does it help?

**Comment by brunoqc** [comment]
No, I still have the problem randomly.

go version go1.4 windows/amd64
gcc (x86_64-posix-seh-rev1, Built by MinGW-W64 project) 4.9.2
SDL2_ttf-devel-2.0.12-mingw
SDL2-devel-2.0.3-mingw

It's weird that I have a different version for ttf but I got the latest version of both yesterday.

**Comment by jackyb** [comment]
Could you show me the error again? Just want to see if it still happens at the same function.

**Comment by brunoqc** [comment]
You're right. I should have seen that.

```
Exception 0xc0000005 0x0 0xffffffffffab4a88 0x6c7be871
PC=0x6c7be871
signal arrived during cgo execution

github.com/veandco/go-sdl2/sdl_ttf._Cfunc_SDL_free(0x38c9d0)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/:44 +0x4c
github.com/veandco/go-sdl2/sdl_ttf.OpenFont(0x520970, 0xf, 0x12c, 0xc082034018, 0x0, 0x0)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/sdl_ttf.go:81 +0x12d
main.main()
        D:/Dev/go/Tests/sdl2crash/sdl2.go:17 +0x73

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        c:/go/src/runtime/asm_amd64.s:2232 +0x1
rax     0xffffffffffab4a70
rbx     0xc082025d10
rcx     0x11afea
rdx     0x0
rdi     0xc082025d10
rsi     0x5bea60
rbp     0x22feb0
rsp     0x22fdf0
r8      0x1000c5ce54cc8bd0
r9      0x1000c5ce5477d640
r10     0x8
r11     0x330000
r12     0x27
r13     0x0
r14     0x0
r15     0x0
rip     0x6c7be871
rflags  0x10212
cs      0x33
fs      0x53
gs      0x2b
```

I added a lock to OpenFont but now I still have :
```
Exception 0xc0000005 0x0 0xffffffffffffffff 0x6c7be90e
PC=0x6c7be90e
signal arrived during cgo execution

github.com/veandco/go-sdl2/sdl_ttf._Cfunc_SDL_free(0x30c9d0)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/:44 +0x4c
github.com/veandco/go-sdl2/sdl_ttf.OpenFont(0x520990, 0xf, 0x12c, 0xc082034018, 0x0, 0x0)
        d:/dev/go/gopath/src/github.com/veandco/go-sdl2/sdl_ttf/sdl_ttf.go:84 +0x14c
main.main()
        D:/Dev/go/Tests/sdl2crash/sdl2.go:17 +0x73

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        c:/go/src/runtime/asm_amd64.s:2232 +0x1
rax     0x30c9c0
rbx     0xc082025d10
rcx     0x30c9d0
rdx     0x1
rdi     0xc082025d10
rsi     0x5bea60
rbp     0x22feb0
rsp     0x22fdf0
r8      0x100079e4665d8228
r9      0x100079e4668e4be8
r10     0x8
r11     0x2f0000
r12     0x27
r13     0x0
r14     0x0
r15     0x0
rip     0x6c7be90e
rflags  0x10297
cs      0x33
fs      0x53
gs      0x2b
```

Line 84 is OpenFont's last line (it was line 81 before I added the lock so it seems the lock isn't working):
```go
return &Font{f}, nil
```

**Comment by akovaski** [comment]
I have the same setup and cannot reproduce this. Perhaps go is building using an old version of go-sdl2; have you tried clearing your %GOROOT%\pkg folder and building again to see if the problem persists?

**Comment by brunoqc** [comment]
@akovaski: Do you mean %GOPATH%\pkg ? If I clear %GOROOT%\pkg I can't build anymore. I get `go tool: no such tool "6g"`.

I still have the problem if I clear %GOROOT%\pkg.

**Comment by akovaski** [comment]
whoops, yeah I meant GOPATH

**Comment by brunoqc** [comment]
I don't have the problem on XP 32-bit with `gcc (i686-posix-dwarf-rev1, Built by MinGW-W64 project) 4.9.2`.

**Comment by brunoqc** [comment]
That's weird. I don't have the problem on 64-bit if I replace all `C.SDL_free` by `C.free`. Even if I remove the locks I added to `sdl_ttf.go`.

`SDL_malloc.c` seems to use a custom malloc (dlmalloc) which is outdated (it doesn't mean it's the cause of my problem). dlmalloc seems to have some options we can define when compiling (like USE_LOCKS).

**Comment by akovaski** [comment]
Ok, I have finally replicated the exception. What triggered when I switched out the SDL2.dll that I compiled from source and put in the SDL2.dll from the SDL-website.

Regarding SDL_free, it seems to me that it should default to the built-in 'free' (at the top of SDL_malloc.c), but perhaps it is using dlmalloc in the website's dll.

In any case, I think SDL_free should only be used with SDL_malloc (and not with C.CString, which I think uses stdlib's malloc) because 'free' functions have to know how the memory is allocated. So, yes, we should switch the freeing of all _CStrings_ from SDL_free to C.free.
